modules/infra/control/port.c

```diff
@@ -626,6 +609,94 @@ static void port_to_api(void *info, const struct iface *iface) {
 	}
 }
 
+static struct event *link_event;
+
+static void link_event_cb(evutil_socket_t, short, void *) {
+	unsigned max_sleep_us, rx_buffer_us;
+	struct rte_eth_rxq_info qinfo;
+	struct rte_eth_link link;
+	struct queue_map *qmap;
+	struct worker *worker;
+	const struct iface *i;
+	struct iface *iface;
+
+	STAILQ_FOREACH (worker, &workers, next) {
+		if (gr_args()->poll_mode)
+			max_sleep_us = 0;
+		else
+			max_sleep_us = 1000; // unreasonably long maximum (1ms)
+
+		arrforeach (qmap, worker->rxqs) {
+			i = port_ifaces[qmap->port_id];
+			if (i == NULL)
+				continue;
+			iface = iface_from_id(i->id);
+			if (iface == NULL)
+				continue;
+
+			if (rte_eth_link_get(qmap->port_id, &link) < 0) {
+				LOG(INFO, "%s: link status down", iface->name);
+				iface->state &= ~GR_IFACE_S_RUNNING;
+				continue;
+			}
+			if (link.link_status == RTE_ETH_LINK_UP) {
+				if (!(iface->state & GR_IFACE_S_RUNNING)) {
+					LOG(INFO, "%s: link status up", iface->name);
+					iface->state |= GR_IFACE_S_RUNNING;
+				}
+			} else {
+				if (iface->state & GR_IFACE_S_RUNNING) {
+					LOG(INFO, "%s: link status down", iface->name);
+					iface->state &= ~GR_IFACE_S_RUNNING;
+				}
+				continue;
+			}
+			if (gr_args()->poll_mode)
+				continue;
+
+			switch (link.link_speed) {
+			case RTE_ETH_SPEED_NUM_NONE:
+			case RTE_ETH_SPEED_NUM_UNKNOWN:
+				continue;
+			}
+			if (rte_eth_rx_queue_info_get(qmap->port_id, qmap->queue_id, &qinfo) < 0)
+				continue;
+
+			rx_buffer_us = queue_buffer_us(link.link_speed, qinfo.nb_desc);
+			if (rx_buffer_us < max_sleep_us)
+				max_sleep_us = rx_buffer_us;
+		}
+		if (atomic_exchange(&worker->max_sleep_us, max_sleep_us) != max_sleep_us)
+			LOG(INFO, "[CPU %u] worker max sleep %uus", worker->cpu_id, max_sleep_us);
+	}
+}
+
+static int lsc_port_cb(uint16_t, enum rte_eth_event_type, void *, void *) {
+	// This callback may be executed from any dataplane or DPDK thread.
```

Comment:
I didn't encounter any case where the callback was called from any other thread. But *in theory*, it could happen as any driver code may call `rte_eth_dev_callback_process()`.